File: LeetCode - Algorithm/Linked List/206.py
# 206. Reverse Linked List
# https://leetcode.com/problems/reverse-linked-list/

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head

        prev = None
        current = head
        next = head.next

        while next:
            current.next = prev

            prev = current
            current = next
            next = next.next

        current.next = prev

        return current

File: LeetCode - Algorithm/Linked List/206.java
// 206. Reverse Linked List
// https://leetcode.com/problems/reverse-linked-list/

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null) {
            return head;
        }

        ListNode prev = null;
        ListNode current = head;
        ListNode next = head.next;

        while (current != null && next != null) {
            current.next = prev;

            prev = current;
            current = next;
            next = next.next;
        }

        current.next = prev;

        return current;
    }
}